src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/DelegateTests.cs

```diff
@@ -0,0 +1,214 @@
+                {
+                    res = funcDelegate (10, 20);
+                }
+                App.call_test_method (""InvokeI32"", [ res, res ]);
+            ");
+
+            Assert.Equal(30, HelperMarshal._functionResultValue);
+            Assert.Equal(60, HelperMarshal._i32Value);
+        }
+
+        [Fact]
+        public static void MarshalFunctionLooptyLoopIncrement()
+        {
+            HelperMarshal._functionResultValue = 0;
+            HelperMarshal._i32Value = 0;
+            Runtime.InvokeJS(@"
+                var funcDelegate = App.call_test_method (""CreateFunctionDelegate"", [  ]);
+                var res = funcDelegate (10, 20);
+                for (x = 0; x < 1000; x++)
+                {
+                    res = funcDelegate (x, x);
+                }
+                App.call_test_method (""InvokeI32"", [ res, res ]);
+            ");
+
+            Assert.Equal(1998, HelperMarshal._functionResultValue);
+            Assert.Equal(3996, HelperMarshal._i32Value);
+        }
+
+        [Fact]
+        public static void MarshalFunctionReturnAction()
+        {
+            HelperMarshal._functionActionResultValue = 0;
+            HelperMarshal._functionActionResultValueOfAction = 0;
+
+            Runtime.InvokeJS(@"
+                var funcDelegate = App.call_test_method (""CreateFunctionDelegateWithAction"", [  ]);
+                var actionDelegate = funcDelegate (10, 20);
+                actionDelegate(30,40);
+            ");
+
+            Assert.Equal(30, HelperMarshal._functionActionResultValue);
+            Assert.Equal(70, HelperMarshal._functionActionResultValueOfAction);
+        }
+
+        [Fact]
+        public static void MarshalActionIntInt()
+        {
+            HelperMarshal._actionResultValue = 0;
+
+            Runtime.InvokeJS(@"
+                var actionDelegate = App.call_test_method (""CreateActionDelegate"", [  ]);
+                actionDelegate(30,40);
+            ");
+
+            Assert.Equal(70, HelperMarshal._actionResultValue);
+        }
+
+        [Fact]
+        public static void MarshalActionFloatIntToIntInt()
+        {
+            HelperMarshal._actionResultValue = 0;
+            Runtime.InvokeJS(@"
+                var actionDelegate = App.call_test_method (""CreateActionDelegate"", [  ]);
+                actionDelegate(3.14,40);
+            ");
+
+            Assert.Equal(43, HelperMarshal._actionResultValue);
+        }
+
+        [Fact]
+        public static void MarshalDelegateMethod()
+        {
+            HelperMarshal._delMethodResultValue = string.Empty;
+            Runtime.InvokeJS(@"
+                var del = App.call_test_method (""CreateDelegateMethod"", [  ]);
+                del(""Hic sunt dracones"");
+            ");
+
+            Assert.Equal("Hic sunt dracones", HelperMarshal._delMethodResultValue);
+        }
+
+        [Fact]
+        public static void MarshalAnonDelegateMethod()
```

Comment:
These kinda tests could be combined like:
```csharp
        [Theory]
        [InlineData("CreateDelegateFromAnonymousMethod_VoidString")]
        [InlineData("CreateDelegateFromLambda_VoidString")]
        [InlineData("CreateDelegateFromMethod_VoidString")]
        [InlineData("CreateActionT_VoidString")]
        public static void InvokeDelegate_VoidString(string creator)
        {
            HelperMarshal._delegateCallResult = string.Empty;
            var s = Runtime.InvokeJS($@"
                var del = App.call_test_method (""{creator}"", [  ]);
                del(""Hic sunt dracones"");
            ");

            Assert.Equal("Notification received for: Hic sunt dracones", HelperMarshal._delegateCallResult);
        }
```

```csharp
        internal static string _delegateCallResult;
        private static Del CreateDelegateFromAnonymousMethod_VoidString()
        {
            // Instantiate the delegate.
            Del handler = delegate(string name) { _delegateCallResult = $"Notification received for: {name}"; };
            return handler;
        }

        private static Del CreateDelegateFromLambda_VoidString()
        {
            // Instantiate the delegate.
            Del handler = (string name) => { _delegateCallResult = $"Notification received for: {name}"; };
            return handler;
        }

        public static void DelegateMethod_VoidString(string name) => _delegateCallResult = $"Notification received for: {name}";

        private static Del CreateDelegateFromMethod_VoidString()
        {
            // Instantiate the delegate.
            Del handler = DelegateMethod_VoidString;
            return handler;
        }

        private static Action<string> CreateActionT_VoidString()
            => (string name) => _delegateCallResult = $"Notification received for: {name}";
```

Reply:
Addressed in recent commits